### ISSUE TITLE
runtime: sigaction problem when using cgo on FreeBSD

### ISSUE THREAD

**Issue by billziss-gh** [golang/go#25745]
### What version of Go are you using (`go version`)?

```
go version go1.10.1 freebsd/amd64
```

### Does this issue reproduce with the latest release?

Likely.

### What operating system and processor architecture are you using (`go env`)?

```
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/billziss/.cache/go-build"
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="freebsd"
GOOS="freebsd"
GOPATH="/home/billziss/go"
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/freebsd_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build089165738=/tmp/go-build -gno-record-gcc-switches"
```

### What did you do?

The following simple cgo program:

```go
package main

/*
#include <stdio.h>
#include <signal.h>

static void sigprint(int sig) {
	struct sigaction sa;
	int ret = sigaction(sig, 0, &sa);
	fprintf(stderr, "ret=%d handler=%p\n", ret, sa.sa_handler);
}
*/
import "C"
import "syscall"

func main() {
	C.sigprint(C.int(syscall.SIGTERM))
}
```

Succeeds (prints a non-0 signal handler) on Darwin and Linux:

```console
$ go run sigbug.go
ret=0 handler=0x4034150
```

But fails (prints a 0 signal handler) on FreeBSD:

```console
$ go run sigbug.go
ret=0 handler=0x0
```

### What did you expect to see?

I expected to see the signal handler that the go runtime installs.

### What did you see instead?

The C `sigaction` function thinks that there is no signal handler.

### Investigation

I believe the reason for the "mysterious" `sigaction` behavior is that the Go runtime invokes [`sigaction`](https://github.com/golang/go/blob/release-branch.go1.10/src/runtime/os_freebsd.go#L310-L319) using [`SYSCALL`](https://github.com/golang/go/blob/release-branch.go1.10/src/runtime/sys_freebsd_amd64.s#L180-L188). However FreeBSD expects that C (and cgo) programs should use the libc [`sigaction`](https://github.com/freebsd/freebsd/blob/release/11.1.0/lib/libc/sys/sigaction.c#L44-L50), which is ["interposed"](https://github.com/freebsd/freebsd/blob/release/11.1.0/lib/libthr/thread/thr_syscalls.c#L668) by libthr to [`__thr_sigaction`](https://github.com/freebsd/freebsd/blob/release/11.1.0/lib/libthr/thread/thr_sig.c#L518). Libthr maintains its own [copy](https://github.com/freebsd/freebsd/blob/release/11.1.0/lib/libthr/thread/thr_sig.c#L58-L65) of `sigaction`'s and [overrides](https://github.com/freebsd/freebsd/blob/release/11.1.0/lib/libthr/thread/thr_sig.c#L567-L572) what the kernel returns.

A likely fix for this is to do what Linux does for `sigaction`. Invoke it using `SYSCALL` if "!cgo" or use the C library `sigaction` if "cgo". See [`rt_sigaction`](https://github.com/golang/go/blob/release-branch.go1.10/src/runtime/cgo_sigaction.go#L20).

Additional information at [billziss-gh/cgofuse#18](https://github.com/billziss-gh/cgofuse/issues/18#issuecomment-390497573).

**Comment by ianlancetaylor** [comment]
I believe that this is already fixed on tip and therefore in the upcoming 1.11 release as part of the work on #14327.  I'm going to close this issue but please do comment if I turn out to be mistaken.

**Comment by billziss-gh** [comment]
Fantastic! I can confirm this is now fixed against the current tip:

```console
$ go version
go version devel +2ce295e954 Wed Jun 6 01:47:31 2018 +0000 freebsd/amd64
$ go run sigbug.go
ret=0 handler=0x44a9c0
```
